---
title: "Single Cell Processing"
author: "Zaki Fadlullah"
date: "05/12/2017"
output: html_document
---

# Pre-processing of sequence data

## Create reference genome

Genome sequence was downloaded from http://ftp.ensembl.org/pub/release-87/fasta/mus_musculus/dna/. GTF file was downloaded from Gencode at ftp://ftp.sanger.ac.uk/pub/gencode/Gencode_mouse/release_M12/gencode.vM12.annotation.gtf.gz.


Since gencode **GTF** file contains `chr`. We need to edit the gencode **GTF** file to contain the same annotation as ensembl reference.  Edit the **GTF** by doing the following ;


1) Remove `chr` 
2) Rename `chrMT` to `chrM`.


```{bash, eval=FALSE}
cat gencode.vM12.annotation.gtf | sed 's/^chrM/chrMT/'  | sed 's/^chr//' > gencode.vM12.annotation_edit.gtf
```


ERCC sequence and gtf files are donwloaded from https://tools.thermofisher.com/content/sfs/manuals/ERCC92.zip. We combine the ERCC **GTF** file with the edited Gencode genome **GTF**. 

```{bash, eval=FALSE}
cat gencode.vM12.annotation_edit.gtf ${ERCCgtf} > ${outDir}/gencode_ERCC.gtf
```


## Merging lane-wise fastq files

For each sample (or single cell), the sample was sequenced across 4 lanes as shown below.

```{r, eval=FALSE}
C208_MEP_neg_L001_R1.fastq.gz
C208_MEP_neg_L002_R1.fastq.gz
C208_MEP_neg_L003_R1.fastq.gz
C208_MEP_neg_L004_R1.fastq.gz

C208_MEP_neg_L001_R2.fastq.gz
C208_MEP_neg_L002_R2.fastq.gz
C208_MEP_neg_L003_R2.fastq.gz
C208_MEP_neg_L004_R2.fastq.gz
```


We merged the lane-wise fastq files for each sample
```{r, eval=FALSE}
cat C208_MEP_neg_L001_R1.fastq.gz C208_MEP_neg_L002_R1.fastq.gz C208_MEP_neg_L003_R1.fastq.gz C208_MEP_neg_L004_R1.fastq.gz > C208_MEP_neg_R1.fastq.gz


cat C208_MEP_neg_L001_R2.fastq.gz C208_MEP_neg_L002_R2.fastq.gz C208_MEP_neg_L003_R2.fastq.gz C208_MEP_neg_L004_R2.fastq.gz > C208_MEP_neg_R2.fastq.gz
```



## Read trimming

We applied read trimming using trimmomatic (version 0.36) with the following settings. 

```{r, eval=FALSE}
{trimmomatic} PE -threads {num_threads} -phred33 {r1_input} {r2_input} {Trimmomatic_Output}_r1_paired.fq {Trimmomatic_Output}_r1_unpaired.fq {Trimmomatic_Output}_r2_paired.fq {Trimmomatic_Output}_r2_upaired.fq ILLUMINACLIP:$TRIMMOMATICDIR/share/adapters/TruSeq3-PE-2.fa:2:30:10 CROP:75 HEADCROP:5 SLIDINGWINDOW:20:20 MINLEN:36
```


The file `TruSeq3-PE-2.fa` was downloaded from https://github.com/timflutre/trimmomatic/blob/master/adapters/TruSeq3-PE-2.fa.


## Mapping

Mapping was performed with STAR (version 2.4.2a). Genome index was build with Genome + ERCC reference combined. 

```{r, eval=FALSE}
STAR --runThreadN 5 \
--runMode genomeGenerate \
--genomeDir ${outIndex} \
--genomeFastaFiles ${REF}/Mus_musculus.GRCm38.dna.chromosome.1.fa ${REF}/Mus_musculus.GRCm38.dna.chromosome.2.fa ${REF}/Mus_musculus.GRCm38.dna.chromosome.3.fa ${REF}/Mus_musculus.GRCm38.dna.chromosome.4.fa ${REF}/Mus_musculus.GRCm38.dna.chromosome.5.fa ${REF}/Mus_musculus.GRCm38.dna.chromosome.6.fa ${REF}/Mus_musculus.GRCm38.dna.chromosome.7.fa ${REF}/Mus_musculus.GRCm38.dna.chromosome.8.fa ${REF}/Mus_musculus.GRCm38.dna.chromosome.9.fa ${REF}/Mus_musculus.GRCm38.dna.chromosome.10.fa ${REF}/Mus_musculus.GRCm38.dna.chromosome.11.fa ${REF}/Mus_musculus.GRCm38.dna.chromosome.12.fa ${REF}/Mus_musculus.GRCm38.dna.chromosome.13.fa ${REF}/Mus_musculus.GRCm38.dna.chromosome.14.fa ${REF}/Mus_musculus.GRCm38.dna.chromosome.15.fa ${REF}/Mus_musculus.GRCm38.dna.chromosome.16.fa ${REF}/Mus_musculus.GRCm38.dna.chromosome.17.fa ${REF}/Mus_musculus.GRCm38.dna.chromosome.18.fa ${REF}/Mus_musculus.GRCm38.dna.chromosome.19.fa ${REF}/Mus_musculus.GRCm38.dna.chromosome.X.fa ${REF}/Mus_musculus.GRCm38.dna.chromosome.Y.fa ${REF}/Mus_musculus.GRCm38.dna.chromosome.MT.fa ${REF_ercc}/ERCC92.fa \
--sjdbGTFfile ${REF}/gencode_ERCC.gtf

```


Mapping was performed with the following parameters. Note we only mapped the "paired" fastq files.  


```{r, eval=FALSE}
STAR --runMode alignReads \
    --runThreadN {num_threads} \
    --genomeLoad LoadAndKeep \
    --genomeDir {genome_dir} \
    --readFilesIn {Trimmomatic_Output}_r1_paired.fq {Trimmomatic_Output}_r2_paired.fq \
    --outSAMtype BAM Unsorted \
    --outFileNamePrefix {STAR_output}
```

## Counting reads

We used ht-seq (version 0.6.1p1) to generate the count table.

```{r, eval=FALSE}
{htseq} --format=bam --stranded=reverse --type=exon --idattr=gene_id {STAR_output}Aligned.out.bam {ref_genome} > {htseq_output}.txt
```


The output of all samples was saved in a file called `raw_read_counts.txt'. This file is also available from GEO with the GEO accession number [GSE107653](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE107653). 

# Donwstream processing

Once we generated the count table using ht-seq, all further processing is conducted in R (**R version 3.4.0 (2017-04-21) -- "You Stupid Darkness"**). The Bioconductor version used is

```{r}
source("http://bioconductor.org/biocLite.R")
```

# Creating scater sceObject 

We created scater object using the count generated by ht-seq. These files are also available in GEO. The following files were downloaded from [GSE107653](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE107653);


```{bash, eval=FALSE}
GSE107653_raw_read_counts.txt
GSE107653_cell_annotation.txt
GSE107653_gene_annotation.txt
```


* GSE107653_raw_read_counts.txt - Contains the read count as obtained by ht-seq  
* GSE107653_cell_annotation.txt - Metadata of the cells  
* GSE107653_gene_annotation.txt - Annotation of the genes (extracted from the gencode GTF file)  

We load the required library
```{r, message=FALSE}
library(scater)
library(ggplot2)
library(ggpubr)
library(dplyr)
library(textclean)
library(knitr)
library(tidyr)
library(DESeq2)
```


We specify directory structure
```{r}
workDir <- "~/JuliaDraper_PLOSgenetics"
dataDir <- file.path(workDir, "data")
outDir <- file.path(workDir, "output")
```


We read the files were downloaded from [GSE107653](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE107653).
```{r}
count <- read.delim(file.path(dataDir, "GSE107653_raw_read_counts.txt"))
cell_anno <-  read.delim(file.path(dataDir, "GSE107653_cell_annotation.txt"),
                         stringsAsFactors = FALSE)
gene_anno <-  read.delim(file.path(dataDir, "GSE107653_gene_annotation.txt"),
                         stringsAsFactors = FALSE)
```
  

We have a look at the count table

```{r}
count[1:5,1:5]
```

Some editing needs to be done on the count table; 

1) The gene id needs to be as row.names
2) The ht-seq information needs to be removed


Changing the gene id to row.names

```{r}
row.names(count)  <- count$X
count <- count[,-1]
count[1:5,1:5]
```


Removing the ht-seq information. This is how the count table looks **BEFORE** removing the ht-seq information
```{r}
tail(count)[1:6,1:3]
```


This is how the count table looks **AFTER** removing the ht-seq information
```{r}
count <-count[-grep("_", row.names(count)), ]
tail(count)[1:6,1:3]
```


Creating phenoData

```{r}
# -------- #
# Creating the metaData (pheno Data) - containing the cell identity mapping percentage...etc
# -------- #
pheno_df <- cell_anno
# Check if the order of the cell in the pheno data is the same as count table
identical(pheno_df$Cell, colnames(count))
# Make the pheno Data
pheno_data <- new("AnnotatedDataFrame", pheno_df)
rownames(pheno_data) <- pheno_data$Cell
```


Creating featureData
```{r}
# -------- #
# Creating the metaData (feature Data) - containing the gene symbol annotation
# -------- #
feature_df <- gene_anno
# Check if the order of genes in the annotation file is the same as count table
identical(feature_df$gene_id, row.names(count))
# Make the featureData
feature_data <- new("AnnotatedDataFrame", feature_df)
rownames(feature_data) <- feature_data$gene_id
```



Finally make the sceSet
```{r}
sce <- scater::newSCESet(
  countData = count,
  phenoData = pheno_data,
  featureData = feature_data)
```


Save the sceSet to disk
```{r}
save(sce, file=file.path(outDir, "sce.Robj"))
```

# Cell and Gene filtering

Lets run the QC
```{r}
sce_1 <- 
  calculateQCMetrics(sce,
                     feature_controls = list(ERCC = fData(sce)$gene_type == "ERCC",
                                             MT= fData(sce)$seqnames == "MT",
                                             Ribosome = fData(sce)$gene_type == "rRNA")
                     )
```


## Cell filtering

Lets plot a few histogram to get the distribution of the data. First we apply a filter based on mapping percentage. 


```{r, fig.height=7, fig.width=10}
# The colour scheme we use
colp <- c("#1F77B4","#AEC7E8","#FF2E0E", "#FFBB78")


df_pData <- pData(sce_1)

p1 <- 
  ggplot(df_pData, aes(x=UnMap_per, fill=Population)) +
  geom_density(alpha=0.6) +
  theme_bw() +
  geom_vline(xintercept = 25, linetype="dashed") +
  scale_fill_manual(
    values=colp)

p2 <- 
    ggplot(df_pData, aes(x=TotalReads, fill=Population)) +
  geom_density(alpha=0.6) +
  theme_bw() +
  geom_vline(xintercept = 250000, linetype="dashed") +
  scale_fill_manual(
    values=colp)

#scater::multiplot(p1,p2, cols=1)

ggarrange(nrow=2, p1,p2, labels = c("A", "B"))

```

From a quick observation of the plots, we will use the following cut-off.


* Remove cells with unmapped reads more than 25%  
* Remove cells with reads less than 250,000  



Second we filter the cells based on other criteria ;


* Number of genes detected  
* Percentage of Mitochondira reads
* Percentage of Ribosome reads
* Percentage of ERCC


Another measure of quality is the proportion of reads mapped to genes in the mitochondrial genome. High proportions are indicative of poor-quality cells ([Ilicic et al, 2016](https://f1000research.com/articles/5-2122/v2#ref-14), [Islam et al., 2014](https://f1000research.com/articles/5-2122/v2#ref-16)), possibly because of increased apoptosis and/or loss of cytoplasmic RNA from lysed cells.

```{r, fig.height=7, fig.width=10}
p1 <- 
  ggplot(df_pData, aes(x=total_features, fill=Population)) +
  geom_density(alpha=0.6) +
  theme_bw()  +
  geom_vline(xintercept = 2500, linetype="dashed") +
  scale_fill_manual(
    values=colp)

p2 <- 
  ggplot(df_pData, aes(x=pct_counts_feature_controls_MT, fill=Population)) +
  geom_density(alpha=0.6) +
  theme_bw() +
  scale_fill_manual(
    values=colp)

p3 <- 
  ggplot(df_pData, aes(x=pct_counts_feature_controls_Ribosome, fill=Population)) +
  geom_density(alpha=0.6) +
  theme_bw() +
  scale_fill_manual(
    values=colp)

p4 <- 
  ggplot(df_pData, aes(x=pct_counts_feature_controls_ERCC, fill=Population)) +
  geom_density(alpha=0.6) +
  theme_bw() +
  geom_vline(xintercept = 15, linetype="dashed") +
  scale_fill_manual(
    values=colp)


#scater::multiplot(p1,p2,p3,p4,cols=2)

ggarrange(p1,p2, p3, p4, labels = c("A", "B", "C", "D"))

```

We note the following observation form the plot 


* A - most cells detected more than 5000 genes
* B - Percentage of mitochondia is about 2-4%. Low percentage is ideal
* C - Simillar percentage or ribosome across cell. Low percentage is ideal
* D - ERCC percentage : The ERCC is not dominating huge number of reads, which is ideal one or two cell have ~100% ERCC


Now lets apply the filtering. We apply the filtering based on these threshold

```{r}
# Threshold used
t_unmap <- 25
t_reads <- 250000
t_features <- 2500
t_MT <- 15
t_Ribo <- 15
t_ERCC <- 15


df_filter <- dplyr::filter(df_pData, 
                           UnMap_per <=t_unmap & 
                             TotalReads >= t_reads &
                             total_features >= t_features &
                             pct_counts_feature_controls_MT <= t_MT &
                             pct_counts_feature_controls_Ribosome <= t_Ribo &
                             pct_counts_feature_controls_ERCC <= t_ERCC)
```


The initial number of cells were **`r nrow(df_pData)`**, after filtering the number becomes **`r nrow(df_filter)`**. Therefore **`r nrow(df_pData) - nrow(df_filter)` **cells were removed

```{r}
sce_Cellfilter <- sce_1[ , sce_1$Cell %in% df_filter$Cell]
```

## Gene filtering

We conduct two stage of gene filtering ;


1) Removing low-abundance genes
2) Remove over-represented genes


### Low abundance

Low-abundance genes are problematic as zero or near-zero counts do not contain enough information for reliable statistical inference. In addition, the discreteness of the counts may interfere with downstream statistical procedures, e.g., by compromising the accuracy of continuous approximations. These genes are likely to be dominated by drop-out events [Brennecke et al., 2013](https://www.nature.com/nmeth/journal/v10/n11/full/nmeth.2645.html), which limits their usefulness in later analyses. Removal of these genes mitigates discreteness and reduces the amount of computational work without major loss of information.


Here, low-abundance genes are defined as ; 


* Those with an average count below a filter threshold of 1.
* Expressed in at less than 3 cells


Filtering by average count
```{r}
ave.counts <- rowMeans(counts(sce_Cellfilter))
keep <- ave.counts >= 1
print(paste0("Before filtering Gene Numbers (including the 0 expression ones): ", dim(fData(sce_Cellfilter))[1]))
print(paste0('After filtering Gene Numbers: ',sum(keep)))
sce_Genefilter <- sce_Cellfilter[keep, ]
```

## Over-represented

Plotting the over-represented genes

We plot it as gene names
```{r}
top_n <- 50
top50 <- head(dplyr::arrange(fData(sce_Genefilter), desc(pct_total_counts)) %>% select(pct_total_counts, gene_id, gene_name), n=top_n)
```

We want to plot the over-represented using thier gene symbole


```{r}
# Use the package (textclean) and mgsub function
top50_sceSet <- sce_Genefilter
featureNames(top50_sceSet) <- 
  mgsub(featureNames(top50_sceSet), top50$gene_id, top50$gene_name)
```


Overrepresented genes
```{r}
plotQC(top50_sceSet, type = "highest-expression", n=50)
```


Another way to look at the data
```{r}
plotFeatureData(sce_Genefilter, aes(x = n_cells_exprs, y = pct_total_counts))
```

One gene is taking about 20% of sequencing reads in all the cells, another gene is taking about 6%. We have a look at this top 3 genes to see what they are. 

```{r}
df_fData <- fData(sce_Genefilter)
knitr::kable(
dplyr::select(df_fData, gene_id, gene_name, gene_type, seqnames, exprs_rank, n_cells_exprs, pct_total_counts) %>%
  dplyr::arrange(desc(exprs_rank)) %>%
  dplyr::top_n(5, exprs_rank)
)
```


We set a threshold of 5% pct_total_conts and remove these two genes

```{r}
keep <- df_fData$pct_total_counts <=5 
print(paste0("Before filtering Gene Numbers: ", dim(fData(sce_Genefilter))[1]))
print(paste0('After filtering Gene Numbers: ',sum(keep)))
sce_Genefilter <- sce_Genefilter[keep, ]
```


# PCA plot

We reproduce some results in the paper. The first is the PCA plot

```{r}
sce <-  sce_Genefilter

p1 <- scater::plotPCA(sce, colour_by="Population", exprs_values="exprs") +
  scale_fill_manual(values = colp)
p1
```

# Collapsing gene id to gene symbol

A good thing to do now is to use the gene symbol as the featureData. We check if there are any duplicated genes
```{r}
df_fData <- fData(sce)
sum(duplicated(df_fData$gene_name))
```


There is `r sum(duplicated(df_fData$gene_name))` duplicated gene symbol. What is this gene?

```{r}
df_fData$gene_name[duplicated(df_fData$gene_name)]
```

Lets have a look at the count of this gene
```{r}
loc_gene <- df_fData$gene_name == "Hspa14"
g1 <- scater::get_exprs(sce, "counts")[loc_gene, ]
df_g <- as.data.frame(g1)
df_g1 <- df_g
df_g1$gene <- row.names(df_g)
dat <- gather(df_g1, gene)
colnames(dat) <- c("gene", "Cell", "value")
ggplot(dat, aes(x=gene, y=value)) +
  geom_boxplot()
```

We remove the ensemble gene with the lower count
```{r}
rowSums(df_g)
```
```{r}
fNames <- featureNames(sce)
loc_f <- fNames == "ENSMUSG00000051396.15"
sce_ori <- sce
sce <- sce_ori[!loc_f,]
featureNames(sce) <- fData(sce)$gene_name
```


We plot the expression of some genes as in the paper.

```{r}
p <- scater::plotPCA(sce, colour_by="Klf1")
p5 <- p$data 
colnames(p5) <- c("PC1", "PC2", "Klf1", "size")
ggplot(p5, aes(x=PC1, y=PC2, colour=Klf1, size=size)) +
  geom_point() +
  scale_colour_gradient(low = "gray88", high = "red1") +
  scale_size_manual(values=3, labels="", name="") +
  theme_bw()

p <- scater::plotPCA(sce, colour_by="Fli1")
p6 <- p$data 
colnames(p6) <- c("PC1", "PC2", "Fli1", "size")
ggplot(p6, aes(x=PC1, y=PC2, colour=Fli1, size=size)) +
  geom_point() +
  scale_colour_gradient(low = "gray88", high = "red1") +
  scale_size_manual(values=3, labels="", name="") +
  theme_bw()

p <- scater::plotPCA(sce, colour_by="Spi1")
p6 <- p$data 
colnames(p6) <- c("PC1", "PC2", "Spi1", "size")
ggplot(p6, aes(x=PC1, y=PC2, colour=Spi1, size=size)) +
  geom_point() +
  scale_colour_gradient(low = "gray88", high = "red1") +
  scale_size_manual(values=3, labels="", name="") +
  theme_bw()
```

# Differential Expression

We perform differential expression with DESeq2.

```{r}
out_m1 <- file.path(outDir, "DESeq2")
if (dir.exists(out_m1)){
  print("Folder exist")
} else {
  dir.create(out_m1)
}
```

We filter genes with more than 75% dropouts.

```{r}
cts <- get_exprs(sce, "counts")
cts_sub <- cts[fData(sce)$pct_total_exprs >= 0.0025,]
# Remove ERCC
cts_sub <- cts_sub[-grep("ERCC", row.names(cts_sub)),]
```

```{r}
print(paste0("Before filtering Gene Numbers: ", nrow(cts)))
print(paste0('After filtering Gene Numbers: ', nrow(cts_sub)))
```


Get some metadata

```{r}
meta_df <- pData(sce)
meta_df <- dplyr::select(meta_df, Population)
```

Creating DEseq object
```{r}
dds <- DESeqDataSetFromMatrix(countData = cts_sub,
                              colData = meta_df,
                              design = ~ Population)
dds <- DESeq(dds)
```


### Extracting results from Runx1_MEP_pos vs Runx1_MEP_neg

```{r}
res <- DESeq2::results(dds, contrast=c("Population","Runx1_MEP_pos","Runx1_MEP_neg"))
resOrdered <- res[order(res$padj),]

write.table(resOrdered, file.path(out_m1, "Runx1_MEP.xls"), quote=F, sep="\t",col.names=NA)

```


### Extracting results from Runx1_CMP_high vs Runx1_CMP_low

```{r}
res <- DESeq2::results(dds, contrast=c("Population","Runx1_CMP_high","Runx1_CMP_low"))
resOrdered <- res[order(res$padj),]

write.table(resOrdered, file.path(out_m1, "Runx1_CMP.xls"), quote=F, sep="\t",col.names=NA)

```


# SessionInfo 
```{r}
sessionInfo()
```